lib/settings_validator.rb

```diff
@@ -0,0 +1,317 @@
+    %i(yml exception_notifier exception_recipients) =>
+      [:validate_value_has_format_email],
+    %i(yml allowed_iri_schemes) => [:validate_value_is_present],
+    %i(yml formality_levels) => [:validate_value_has_keys, :name, :description],
+    %i(yml license_models) => [:validate_value_has_keys, :name, :url],
+    %i(yml ontology_types) =>
+      [:validate_value_has_keys, :name, :description, :documentation],
+    %i(yml tasks) => [:validate_value_has_keys, :name, :description],
+
+    # hets.yml
+    %i(yml hets_path) => [:validate_value_is_absolute_filepath],
+    %i(yml hets_lib) => [:validate_value_is_absolute_filepath],
+    %i(yml hets_owl_tools) => [:validate_value_is_absolute_filepath],
+    %i(yml cmd_line_options) => [:validate_value_is_present],
+    %i(yml server_options) => [:validate_value_is_present],
+  }
+
+  attr_reader :errors
+
+  def initialize(config)
+    @config = config
+  end
+
+  def validate
+    @errors = []
+
+    HAS_TO_BE_PRESENT.each { |key_chain| validate_presence(key_chain) }
+    HAS_TO_BE_URL.each { |key_chain| validate_format_url(key_chain) }
+    HAS_TO_BE_EMAIL.each { |key_chain| validate_format_email(key_chain) }
+    HAS_TO_HAVE_TYPE.each do |types, key_chains|
+      key_chains.each { |key_chain| validate_type(key_chain, types) }
+    end
+
+    if Rails.env.production?
+      HAS_TO_BE_DIRECTORY_IN_PRODUCTION.each do |key_chain|
+        validate_directory_exists(key_chain)
+      end
+      HAS_TO_BE_PRESENT_IN_PRODUCTION.each do |key_chain|
+        validate_presence(key_chain)
+      end
+    end
+
+    validate_arrays
+    validate_specials
+
+    if @errors.present?
+      raise ValidationError.new(@errors.map(&:message).join("\n"))
+    end
+  end
+
+  protected
+
+  def validate_arrays
+    ARRAY_VALIDATIONS.each do |key_chain, call|
+      if array = get_value(key_chain)
+        array.each do |value|
+          method = call.first
+          args = call[1..-1]
+          send(method, key_chain, value, *args)
+        end
+      end
+    end
+  end
+
+  def validate_specials
+    validate_hets_path
+  end
+
+  def validate_hets_path
+    key_chain = %i(yml hets_path)
+    hets_paths = get_value(key_chain)
+    unless hets_paths.any? { |path| File.executable?(path) }
+      @errors << ResourceNotFound.new(key_chain,
+        "Executable 'hets' not found in any of the given paths: #{hets_paths}")
+    end
+  end
+
+  def get_value(key_chain)
+    object = base(key_chain.first)
+    key_chain[1..-1].each do |key|
+      object = object.send(key)
+      if object.nil?
+        @errors << KeyNotSet.new(key_chain, 'Key not set.')
+        return
+      end
+    end
+    object
+  end
+
+  def base(namespace)
+    if namespace == :yml
+      Settings
+    else
+      @config
+    end
+  end
+
+  def validate_with(key_chain, &block)
+    if value = get_value(key_chain)
+      block.call(value)
+    end
+  end
+
+  def validate_directory_exists(key_chain)
```

Comment:
a) skip meta - doesn't change anything. As a starting point [wikipedia] (http://en.wikipedia.org/wiki/XML_schema) might be sufficient.

b) No, not at all! Again, reasonable defaults are a application internal thing! When a usual user can change it to whatever he wants, why should they still be considered reasonable. When you say *settings* you actually give away the control completely and can't guarantee anything! The only public place, where they should appear is in the documentation! BTW, you flipflop a lot - one time you say, overwrite by a 2nd file is "not clean", the next time you say, one should overwrite using a 2nd file. Looks like a big confusion to me.

c) You see, re-inventing the wheel doesn't make sense unless it is really worth the burden - in the industry they would ask you: show me your business case. If you than tell the manager, that you just need to check 1-2 dozen of app parameters, he'll say, what? Because you are to lazy to write down ~3 simple LOCs/param you wanna create another many man-month project, which probably no one will use, because writing down the description for the parameters takes at least the same time and probably not much less LOCs than when just doing it the usual, simple way? "Kanonenkugeln und Spatzen ..."

c,d) The point is: your class covers the very basic things, ~10% , 90% are still missing. So you made the application more complex and harder to understand but failed to actually solv the problem. An example for e.g. min max is hets workers.

e) Sorry, but reading and validating data read should be one step. The paradigm which goes hand in hand with this is "fail fast"! And please, keep focused on the task. I didn't say that's the right place do it, I said, that's what you could have been used (and implicit what experienced programmers would have been done) without wasting a lot of your and _my_ time! Also I get tiered about your "that's how rails tells one should do". I don't believe this non-sense anymore - so if you make such statements it would be nice to give a reference for this.

f) data validation is always a boring business, but needs to be done and wrt. to app parameters  it is damn easy. More or less a copy and paste thing - I say for the most cases a blind janitor could do that. So I could not understand at all, why you make a mountain out of a molehill ...

g) But enough to introduce other bugs ...

BTW: That's it for me. I supplied proper/workable suggestions and if you don't wanna take it, make what you want. If you want it to be stupid, keep it stupid. I'm scared of your NIH attitude.